### PR TITLE
runMonitoring take more types addition

### LINKED ISSUE #2380: Make `runMonitoring` take more types
Currently `runMonitoring` only takes job slices. Would be nice to have it accept an `int`, or a list of `int` or a `job` object.

### PR DESCRIPTION
**Pull request by saadkhi**
The issue solved was to make `runMonitoring` accept different job inputs—specifically, an integer (job ID), a list of integers (job IDs), and a job object—in addition to the original job slice format. This increases flexibility in specifying jobs for monitoring.

fixed #2380

### CODE REVIEW

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 1]
```diff
@@ -706,110 +706,107 @@ def reloadJob(self, i):
         stripProxy(self.registry_slice).objects.repository.load([i])
         return True
 
-    def runMonitoring(self, jobs=None, steps=1, timeout=300):
+    def runMonitoring(self, steps, timeout, jobs=None):
```
Comment (mesmith75) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py]: We don't want to change these arguments

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 2]
```diff
@@ -706,110 +706,107 @@ def reloadJob(self, i):
...
-          jobs: a registry slice to be monitored (None -> all jobs), it may be passed by the user so ._impl is stripped if needed
+        steps:   number of monitoring steps to run
+        timeout: how long to wait for monitor steps termination (seconds)
+        jobs: a registry slice to be monitored (None -> all jobs), or an int, list of int, or job object
+        
         Return:
-          False, if the loop cannot be started or the timeout occured while waiting for monitoring termination
-          True, if the monitoring steps were successfully executed  
-        Note:         
-          This method is meant to be used in Ganga scripts to request monitoring on demand.
```
Comment (mesmith75) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py:720]: Could you leave this part of the docstring?

**Comment by mesmith75** [comment]
You seem to have included your virtualenv in the commit - could you remove that?

You also seem to have touched a lot of things that are best left alone. Also I think some issues have occurred with indenting/whitespace etc. Could you try and tidy that up?

The issue only needs some pretty minimal changes so could you try and tidy this up to only include the minimal adjustments necessary?

**Comment by saadkhi** [comment]
Yes, I will make changes and update soon.

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 3]
```diff
@@ -757,59 +756,63 @@ def runMonitoring(self, jobs=None, steps=1, timeout=300):
...
+                m_jobs = self.registry_slice(jobs)
+            elif isinstance(jobs, list):
+                # If jobs is a list of int, ensure all elements are int and convert to registry slice
+                if all(isinstance(job, int) for job in jobs):
+                    log.debug(f"Converting job IDs {jobs} to a registry slice")
+                    m_jobs = [self.registry_slice(job) for job in jobs]
+                else:
+                    log.warning("List must contain integers representing job IDs")
+                    return False
+            elif isinstance(jobs, jobs):
```
Comment (egede) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py]: This seems to make no sense to me. Is there an overload of symbols here (between the argument jobs to the function and the global jobs object?

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 4]
```diff
@@ -757,59 +756,63 @@ def runMonitoring(self, jobs=None, steps=1, timeout=300):
...
+            #log.debug("jobs: %s" % str(jobs))
+            #log.debug("self.__mainLoopCond: %s" % str(self.__mainLoopCond))
+
+        # Handle the `jobs` input (new logic to handle int, list, job object)
+        if jobs is not None:
+            if isinstance(jobs, int):
+                # If jobs is a single int, convert it to a slice (equivalent to one job)
+                log.debug(f"Converting job ID {jobs} to a registry slice")
+                m_jobs = self.registry_slice(jobs)
+            elif isinstance(jobs, list):
```
Comment (egede) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py:768]: Should a list of job objects be allowed as well? Like
```
j1 = Job()
j2 = Job()
runMonitoring([j1,j2])
```

**Review by egede — CHANGES_REQUESTED** [review]

**Comment by egede** [comment]
There is something odd going on with the testing here. Not sure what as it seems unrelated to the code changes that you have made.

**Comment by alexanderrichards** [comment]
@saadkhi I think for the failure of the autopep8 test, when you created the pull request did you click the checkbox to allow upstream developers to make changes to your branch? My best guess at the moment for whats happening here is it's trying to create the autopep8 branch to pull against *your* branch `Enh_int`

**Comment by egede** [comment]
@saadkhi Take a look at the GangaCore Integration tests. There are 3 errors there that relate to the code that you modified. Please take a look first yourself and then ask for help if you can't figure out what is wrong. 

Check the [full details](https://github.com/ganga-devs/ganga/actions/runs/11601147069/job/32466884666?pr=2390). Extract below.

```
FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_b_EnableMonitoring - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_g_runMonitoring_withJobIDList - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_h_runMonitoring_withJobObject - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
```

**Comment by saadkhi** [comment]
> @saadkhi I think for the failure of the autopep8 test, when you created the pull request did you click the checkbox to allow upstream developers to make changes to your branch? My best guess at the moment for whats happening here is it's trying to create the autopep8 branch to pull against _your_ branch `Enh_int`

I'm not sure about it, let me check and fix what is going wrong with it.

**Comment by saadkhi** [comment]
> @saadkhi Take a look at the GangaCore Integration tests. There are 3 errors there that relate to the code that you modified. Please take a look first yourself and then ask for help if you can't figure out what is wrong.
> 
> Check the [full details](https://github.com/ganga-devs/ganga/actions/runs/11601147069/job/32466884666?pr=2390). Extract below.
> 
> ```
> FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_b_EnableMonitoring - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
> FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_g_runMonitoring_withJobIDList - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
> FAILED ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py::TestMonitoring::test_h_runMonitoring_withJobObject - TypeError: AsyncMonitoringService.runMonitoring() got an unexpected keyword argument 'steps'
> ```

yes, @egede I'm working on it to figure out why these test cases get failling.

**Comment by saadkhi** [comment]
> @saadkhi I think for the failure of the autopep8 test, when you created the pull request did you click the checkbox to allow upstream developers to make changes to your branch? My best guess at the moment for whats happening here is it's trying to create the autopep8 branch to pull against _your_ branch `Enh_int`

I’m having trouble understanding what autopep8 is and how it applies to my pull request. Could you please explain it, or if possible, arrange a meeting sometime this week to discuss?

**Comment by alexanderrichards** [comment]
> I’m having trouble understanding what autopep8 is and how it applies to my pull request. Could you please explain it, or if possible, arrange a meeting sometime this week to discuss?

so PEP8 is the python style guidelines for coding. Autopep8 is a tool that when run over a codebase will convert all files to the PEP8 standard. As part of our CI/CD pipeline we run an action that automatically runs this autopep8 tool over the files that you have modified and creates a new branch/pull request with the changes necessary to meet the PEP8 standard.

Hopefully that is clear. If not, I can have a separate conversation with you or alternatively you could probably join our Monday morning (UK time) meeting.

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 5]
```diff
@@ -757,59 +756,59 @@ def runMonitoring(self, jobs=None, steps=1, timeout=300):
...
+                    return False
+            else:
+                # Handle job slices (existing behavior)
+                from GangaCore.GPIDev.Lib.Registry.RegistrySlice import RegistrySlice
+                if not isinstance(jobs, RegistrySlice):
+                    log.warning('jobs argument must be a registry slice, int, list of int, or job object')
+                    return False
+                m_jobs = jobs
+            # Pass the new `m_jobs` (registry slice) for further processing
+            self.makeUpdateJobStatusFunction(jobSlice=m_jobs)
```
Comment (mesmith75) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py]: Why has the order changed? I think we want to grab the lock first still

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py [hunk 6]
```diff
@@ -757,59 +756,59 @@ def runMonitoring(self, jobs=None, steps=1, timeout=300):
...
-
-                #self.registry_slice = m_jobs
-                #log.debug("m_jobs: %s" % str(m_jobs))
-                self.makeUpdateJobStatusFunction(jobSlice=m_jobs)
-
-            log.debug("Enable Loop, Clear Iterators and setCallbackHook")
-            # enable mon loop
+            log.debug("Enable Loop, Clear Iterators, and setCallbackHook")
             self.enabled = True
-            # set how many steps to run
```
Comment (mesmith75) [inline ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py:788]: Why are we losing all these comments?

File: ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py [hunk 7]
```diff
@@ -46,7 +46,9 @@ def test_b_EnableMonitoring(self):
 
         dummySleep(j)
 
-        self.assertNotEqual(jobs(0).status, 'submitted')
```
Comment (mesmith75) [inline ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py]: I don't think we want to change this test.

**Comment by saadkhi** [comment]
> > I’m having trouble understanding what autopep8 is and how it applies to my pull request. Could you please explain it, or if possible, arrange a meeting sometime this week to discuss?
> 
> so PEP8 is the python style guidelines for coding. Autopep8 is a tool that when run over a codebase will convert all files to the PEP8 standard. As part of our CI/CD pipeline we run an action that automatically runs this autopep8 tool over the files that you have modified and creates a new branch/pull request with the changes necessary to meet the PEP8 standard.
> 
> Hopefully that is clear. If not, I can have a separate conversation with you or alternatively you could probably join our Monday morning (UK time) meeting.

Okay, I'll join the meeting and then will solve the autopep8 issue after meeting as soon as possible.

**Comment by alexanderrichards** [comment]
> Okay, I'll join the meeting and then will solve the autopep8 issue after meeting as soon as possible.

👍🏻 The meeting is in 17 mins. Do you have the connection details? If you are on the mattermost channel there is a link at the top

File: ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py (hunk 5 above)
Comment (saadkhi) [reply ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py]: What I have to do, kindly explain. I’ll update it soon.

Comment (mesmith75) [reply ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py]: You have moved this entire chunk of code - I think it needs to go back to its original location

**Review by egede — COMMENTED** [review]
There is something really fishy here. 

It turns out that we have two `runMonitoring` functions in the codebase. One in [MonitoringService.py](https://github.com/ganga-devs/ganga/blob/35e3bcf552519c8aae5f36c8a5d6cef49255f68d/ganga/GangaCore/Core/MonitoringComponent/MonitoringService.py#L235) and one in [Local_GangaMC_Service.py](https://github.com/ganga-devs/ganga/blob/35e3bcf552519c8aae5f36c8a5d6cef49255f68d/ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py#L709). 

If I force the runMonitoring method that is edited in this PR to always return False, then the tests are still passing! My conclusion is that the code is never called.

File: ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py (hunk 7 above)
Comment (alexanderrichards) [reply ganga/GangaCore/test/GPI/Monitoring/TestMonitoring.py]: I agree, we shouldn't be changing the testing files at all (except to add new tests)

**Comment by egede** [comment]
So looking in the [coverage of the testing](https://github.com/ganga-devs/ganga/actions/runs/11670321659/job/32531509439?pr=2390#step:5:473) I indeed see that the code that you have implemented is never executed. So in fact the entire `runMonitoring` function in [Local_GangaMC_Service.py](https://github.com/ganga-devs/ganga/blob/35e3bcf552519c8aae5f36c8a5d6cef49255f68d/ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py#L709) can be removed.

**Comment by egede** [comment]
So the real question now is what the new test files are actually testing. At the moment they just look for `runMonitoring` returning `True`, but they should be extended to see if the jobs in question are actually monitored (and others not). It could be the argument is just ignored. Sorry @saadkhi but this is turning more complicated that we realised and we probably gave you the wrong advise to begin with.

**Comment by saadkhi** [comment]
> So the real question now is what the new test files are actually testing. At the moment they just look for `runMonitoring` returning `True`, but they should be extended to see if the jobs in question are actually monitored (and others not). It could be the argument is just ignored. Sorry @saadkhi but this is turning more complicated that we realised and we probably gave you the wrong advise to begin with.


Hi @egede, thank you for your guidance. Initially, I created the test cases based on the `Local_GangaMC_Service` implementation in [Ganga](https://github.com/ganga-devs/ganga/blob/35e3bcf552519c8aae5f36c8a5d6cef49255f68d/ganga/GangaCore/Core/MonitoringComponent/Local_GangaMC_Service.py#L709), but encountered errors during your testing process. After that, I noticed the `runMonitoring` functionality in the [MonitoringService](https://github.com/ganga-devs/ganga/blob/35e3bcf552519c8aae5f36c8a5d6cef49255f68d/ganga/GangaCore/Core/MonitoringComponent/MonitoringService.py#L235) and adjusted the test cases accordingly. One more things, just tell that the issue is from my side thats why code is not merged.

I’ll continue to explore the Ganga project and improve my contributions alongside the team. I appreciate your support and feedback. Thank you again!

**Comment by mesmith75** [comment]
I think this went a bit off-piste, so closing.

**Pull request closed without merging**